Привет, Дмитрий!

Меня зовут Вячеслав Погорелый, и я буду проводить ревью твоего проекта. 

Предлагаю перейти на "ты", если не против)

Но если это не удобно - дай знать, и мы перейдем на "вы". Моя главная задача — не находить сделанные тобою ошибки, а передать свой опыт и помочь тебе стать аналитиком данных.

Я отметил комментарии и замечания разным цветом далее по тексту markdown:

---    
<div class="alert alert-block alert-danger">
    
### Ошибка
    
Критическое замечание, которые следует исправить.
</div> 
    
---
    
<div class="alert alert-block alert-warning">
    
### Совет
    
Замечание\совет на будущее.
</div>
    
---
    
<div class="alert alert-block alert-success">
    
### Успех

Все здорово

<div class="alert alert-block alert-success">
 
# Итоговый Комментарий ревьюера ver.2
    
#### Успех

Спасибо тебе за оперативные доработки проекта! Вот а теперь я его готов принять. От себя я хочу пожелать успехов в дальнешей учебе.  Ты точно справишься и дальше, я уверен! 

Проект я принимаю и желаю тебе успехов в дальнейшей учебе. С Наступающим, конечно же:)

# Напишите заголовок проекта здесь

- Автор: Галыгин Дмитрий
- Дат: 28.12.2025

### Цели и задачи проекта

#### ЦЕЛЬ: Готовые данные для подготовки статьи, котрую будет готовить комнады игры "Секреты теснолесья".
#### ЗАДАЧА: Обработатьи и выдать "чистые" данные по запросу команды для того, чтобы их можно было удобно использовать в написании статьи.


### Описание данных

Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

- Name — название игры.
- Platform — название платформы.
- Year of Release — год выпуска игры.
- Genre — жанр игры.
- NA sales — продажи в Северной Америке (в миллионах проданных копий).
- EU sales — продажи в Европе (в миллионах проданных копий).
- JP sales — продажи в Японии (в миллионах проданных копий).
- Other sales — продажи в других странах (в миллионах проданных копий).
- Critic Score — оценка критиков (от 0 до 100).
- User Score — оценка пользователей (от 0 до 10).
- Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

<font color='#777778'>Перечислите основные шаги проекта или напишите оглавление. Используйте описание проекта, чтобы зафиксировать основные шаги.</font>

1. Загрузка и знакомство с данными
2. Проверка ошибок в данных и их предобработка
3. Фильтрация данных
4. Категоризация данных

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
df.info() #Вывлеи общую информцию о данных

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
df.isna().sum() #Считаем количество пропусков в каждом столбце

Name                  2
Platform              0
Year of Release     275
Genre                 2
NA sales              0
EU sales              0
JP sales              0
Other sales           0
Critic Score       8714
User Score         6804
Rating             6871
dtype: int64

In [5]:
df.isna().sum()/len(df)*100 #Расчитываем количество пропусокв в % соотношении от обшего количества строк

Name                0.011795
Platform            0.000000
Year of Release     1.621845
Genre               0.011795
NA sales            0.000000
EU sales            0.000000
JP sales            0.000000
Other sales         0.000000
Critic Score       51.391838
User Score         40.127389
Rating             40.522529
dtype: float64

In [104]:
df.head()

name platform year_of_release         genre   na_sales  \
0                wii sports      WII      2006-01-01        sports  41.360001   
1         super mario bros.      NES      1985-01-01      platform  29.080000   
2            mario kart wii      WII      2008-01-01        racing  15.680000   
3         wii sports resort      WII      2009-01-01        sports  15.610000   
4  pokemon red/pokemon blue       GB      1996-01-01  role-playing  11.270000   

    eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0  28.959999      3.77         8.45          76.0         8.0      E  
1   3.580000      6.81         0.77           0.0         0.0    NaN  
2  12.760000      3.79         3.29          82.0         8.3      E  
3  10.930000      3.28         2.95          80.0         8.0      E  
4   8.890000     10.22         1.00           0.0         0.0    NaN

In [7]:
df.nunique() #посчитали количество уникальных значений в столбцах

Name               11559
Platform              31
Year of Release       37
Genre                 24
NA sales             402
EU sales             308
JP sales             245
Other sales          155
Critic Score          82
User Score            96
Rating                 8
dtype: int64

In [8]:
df.duplicated(subset = ['Name'], keep = 'last').sum() #Вычислим количестов дубликатов только в стобце 'Name', кроме первых упоминаний этих дубликатов в столбце

5396

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<font color='#777778'>Используйте ячейки типа Markdown для промежуточных выводов и расширенных комментариев к действиям с данными. </font>

### После первичного анализа можно сделать следующие выводы:
1. Названия столбов некорректные
2. Присутствуют пропуски в столбцах, с котоырми нужно порабоатть
3. Некорректные типы данных в столбцах
4. есть явные дубликаты,  котрые необходимо удалить

## Действия которые необходимо сделать для отчистки и подготовки данных:
1. Изменить название столбов, чтобы они соответствовали стилю snake case.


2. Откорректировать тип данных в столбцах, например: 
- данные про год релиза приветсик  типу datetime64
- данные по продажам и оценке привести к типу float64
3. По пропускам:
- 2 пропуска в названи игры- явная ошибка, сроки можно удалить
- есть пропуски в столбце по году выпуска  Year of Release, они составляют 1,6% от общего кол-ва строк, значение не существенное для анализа, можно оставить
- много пропусков в оценеке и рейтинге, их появление понятно- возможно просто нет оценки, но я бы их удалил, так как наша основная задача проанализировать и проранжировать игры по этим критериям.
4. Дубликаты:
- исходя их информации в столбцах, наличие дубликтаов в столбцах адекватно везде, кроме первого - Name, это явные дбликаты, так как не может быть несколько одинаковых названий. Их нужно удалить.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [9]:
print(df.columns) #Вывели все названия столбцов

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')


In [114]:
df.columns = df.columns.str.lower() #Поменяли регистр столбцов на нижний

In [115]:
df.columns = df.columns.str.replace(' ', '_') #Заменили все пропуски на нижние подчеркивания
print(df.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [116]:
#Переведем данные в столбцах с оценкой и продажами в формат float и применими параметр downcast для оптимизация размера данных. 
#Выбран тип float потому что данные имеют не целые числа и в дальнейшем анализе не хотелось бы их упускать
for column in ['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']:
    df[column] = df[column].fillna(0)
    
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11559 entries, 0 to 16953
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   name             11559 non-null  object        
 1   platform         11559 non-null  object        
 2   year_of_release  11395 non-null  datetime64[ns]
 3   genre            11559 non-null  object        
 4   na_sales         11559 non-null  float32       
 5   eu_sales         11559 non-null  float32       
 6   jp_sales         11559 non-null  float32       
 7   other_sales      11559 non-null  float32       
 8   critic_score     11559 non-null  float32       
 9   user_score       11559 non-null  float32       
 10  rating           5941 non-null   object        
dtypes: datetime64[ns](1), float32(6), object(4)
memory usage: 812.7+ KB
None


In [117]:
df['year_of_release'] = pd.to_datetime(df['year_of_release'], format = '%Y') #Преобразуем данные в столбце year_of_release в формат datetime и укажем параметр %y
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11559 entries, 0 to 16953
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   name             11559 non-null  object        
 1   platform         11559 non-null  object        
 2   year_of_release  11395 non-null  datetime64[ns]
 3   genre            11559 non-null  object        
 4   na_sales         11559 non-null  float32       
 5   eu_sales         11559 non-null  float32       
 6   jp_sales         11559 non-null  float32       
 7   other_sales      11559 non-null  float32       
 8   critic_score     11559 non-null  float32       
 9   user_score       11559 non-null  float32       
 10  rating           5941 non-null   object        
dtypes: datetime64[ns](1), float32(6), object(4)
memory usage: 812.7+ KB
None


- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [118]:
#Переведем данные в столбцах с оценкой и продажами в формат float и применими параметр downcast для оптимизация размера данных. 
#Выбран тип float потому что данные имеют не целые числа и в дальнейшем анализе не хотелось бы их упускать
for column in ['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']:
    df[column] = pd.to_numeric(df[column], errors = 'coerce', downcast = 'float')
    
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11559 entries, 0 to 16953
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   name             11559 non-null  object        
 1   platform         11559 non-null  object        
 2   year_of_release  11395 non-null  datetime64[ns]
 3   genre            11559 non-null  object        
 4   na_sales         11559 non-null  float32       
 5   eu_sales         11559 non-null  float32       
 6   jp_sales         11559 non-null  float32       
 7   other_sales      11559 non-null  float32       
 8   critic_score     11559 non-null  float32       
 9   user_score       11559 non-null  float32       
 10  rating           5941 non-null   object        
dtypes: datetime64[ns](1), float32(6), object(4)
memory usage: 812.7+ KB
None


### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [119]:
df.isna().sum() #Абсолютное количество пропусков

name                  0
platform              0
year_of_release     164
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score          0
user_score            0
rating             5618
dtype: int64

In [120]:
df.isna().sum()/len(df)*100 #Отнсоительное количество пропусков

name                0.000000
platform            0.000000
year_of_release     1.418808
genre               0.000000
na_sales            0.000000
eu_sales            0.000000
jp_sales            0.000000
other_sales         0.000000
critic_score        0.000000
user_score          0.000000
rating             48.602820
dtype: float64

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


### Оценка пропусков:
1. По количеству пропусков в Name и Genre можно сказать, что это ошибка и их можно удалить
2. В году выпуска 1,6% пропуска игр, их можно оставить:
    - значение не критичное для анализа 
    - при сортировке по годам они и так не войдут в анализ
3. В user_score и rating пропуски можно объянить тем, что игроки не оставили свои оценки и организация  ESRB не дала своей оценки, это вполне возможно, поэтому я бы их оставил
4. есть пропуски в продаже игры, в столбцах eu_sales, jp_sales. Их можно заменить на усреденные значения.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [121]:
df= df.dropna(subset = ['name']) #Удалили пропуски в Name и в genre
print(df.isna().sum())

name                  0
platform              0
year_of_release     164
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score          0
user_score            0
rating             5618
dtype: int64


In [122]:
eu_sales_min = df['eu_sales'].min()
eu_sales_max = df['eu_sales'].max()
eu_sales_mean = df['eu_sales'].mean()
eu_sales_median = df['eu_sales'].median()
jp_sales_min = df['jp_sales'].min()
jp_sales_max = df['jp_sales'].max()
jp_sales_mean = df['jp_sales'].mean()
eu_sales_median = df['eu_sales'].median()

print(eu_sales_min, eu_sales_max, eu_sales_mean, eu_sales_median)
print(jp_sales_min, jp_sales_max, jp_sales_mean, eu_sales_median)
#Разброс в значених очень большой, при этом медиана и среднее значение существенно ближе к нижней границе значений. 
#Согласен с замечанием выше о том, что их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

0.0 28.96 0.15610218 0.02
0.0 10.22 0.10413518 0.02


In [123]:
df['eu_sales'] = df['eu_sales'].fillna(df.groupby(['platform', 'year_of_release'])['eu_sales'].transform('mean'))
df['jp_sales'] = df['jp_sales'].fillna(df.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean'))
print(df.info())
#В этом случае заменили на среднее значение в рамках категории платформы и года реализации

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11559 entries, 0 to 16953
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   name             11559 non-null  object        
 1   platform         11559 non-null  object        
 2   year_of_release  11395 non-null  datetime64[ns]
 3   genre            11559 non-null  object        
 4   na_sales         11559 non-null  float32       
 5   eu_sales         11559 non-null  float32       
 6   jp_sales         11559 non-null  float32       
 7   other_sales      11559 non-null  float32       
 8   critic_score     11559 non-null  float32       
 9   user_score       11559 non-null  float32       
 10  rating           5941 non-null   object        
dtypes: datetime64[ns](1), float32(6), object(4)
memory usage: 812.7+ KB
None


### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

Как было писано в первом пункте есть множество дубликатов в столбце с названием игры - 5396, сейчас будем работать с нмии.
в остальных столбцах дубликаты уместны, так как оценка, год и жанр могут повторяться и совпадать.

In [124]:
unique_genres = df['genre'].unique()
print(unique_genres)
#В жанрах есть дубликаты связанные с разным регситром, приведм все к нижнему регситру

['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'fighting' 'adventure' 'strategy']


In [125]:
unique_platforms = df['platform'].unique()
print(unique_platforms)
#В столбце с платформами нет дубликатов связанных с разным форматом

['WII' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'WIIU' 'GC' 'GEN' 'XONE' 'DC' 'SAT' 'PSV'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']


Приведем столбы name и genre  к нижнему регистру

In [129]:
df['name'] = df['name'].str.lower()
df['genre'] = df['genre'].str.lower()

In [130]:
#Удалим дубликаты в столбце name 
df.drop_duplicates(subset = ['name'], inplace = True)

In [131]:
df.duplicated(subset = ['name'], keep = 'last').sum()

0

In [132]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11559 entries, 0 to 16953
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   name             11559 non-null  object        
 1   platform         11559 non-null  object        
 2   year_of_release  11395 non-null  datetime64[ns]
 3   genre            11559 non-null  object        
 4   na_sales         11559 non-null  float32       
 5   eu_sales         11559 non-null  float32       
 6   jp_sales         11559 non-null  float32       
 7   other_sales      11559 non-null  float32       
 8   critic_score     11559 non-null  float32       
 9   user_score       11559 non-null  float32       
 10  rating           5941 non-null   object        
dtypes: datetime64[ns](1), float32(6), object(4)
memory usage: 812.7+ KB
None


- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [133]:
print(df.duplicated().sum())

0


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

Повторов в датафрейме больше нет, все строки являются уникальными. Было обнаружено 5396 дулбикатов.
Мы удалил дубликаты в столбце с назвагнием игры, так как не может быть две игры с одним и тем же названием

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [134]:
count_dublicated = 16956 - df.shape[0]
perc_count_dublicated = round(count_dublicated/16956*100, 2)
print(f"Количесвто удаленных дубликатов в абсолютном занчении: {count_dublicated}")
print(f"Количесвто удаленных дубликатов в относительном занчении: {perc_count_dublicated}")

Количесвто удаленных дубликатов в абсолютном занчении: 5397
Количесвто удаленных дубликатов в относительном занчении: 31.83


- После проведения предобработки данных напишите общий промежуточный вывод.

### Я провел предобработку данных, были выполнены следующие действия:
1. Изменил название столбцов, чтобы они соответствовали стилю snake case.
2. исправили типы даннных в столбцах:
 - данные в столбце год релиза привели к типу datetime64
 - данные по продажам и оценке привели к типу float64
3. Поработал с пропусками:
 - Удалил пропуски в Name и в genre
 - пропуски св столбцах поп радажам заменили на среднее значение по ктаегориям жанра и года реализации игры
 - пропуски в году выпуска игры решил оставить, дальнейшая фильтрация исключит их
 - все строковые значения в числовых столбцах привели к значению Nan
4. Дубликаты:
 - выявил явные и неявные дубликаты в стобца name и genre
 - неявные дубликаты в стоолбце genre убрали с помощью замены регистра на нижний 
 - удалили строки с дубликатами в столбце name, их было 5397 или 31.8%

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [135]:
df_actual = df[(df['year_of_release'] >= '2000-01-01') &
               (df['year_of_release'] <= '2013-12-31')]
print(df_actual)
#Фильтруем данные по году выпуска между 2000 и 2013 гоадми

                                     name platform year_of_release  \
0                              wii sports      WII      2006-01-01   
2                          mario kart wii      WII      2008-01-01   
3                       wii sports resort      WII      2009-01-01   
6                   new super mario bros.       DS      2006-01-01   
7                                wii play      WII      2006-01-01   
...                                   ...      ...             ...   
16943             storm: frontline nation       PC      2011-01-01   
16945                            plushees       DS      2008-01-01   
16946                             15 days       PC      2009-01-01   
16949  woody woodpecker in crazy castle 5      GBA      2002-01-01   
16952                    lma manager 2007     X360      2006-01-01   

            genre   na_sales   eu_sales  jp_sales  other_sales  critic_score  \
0          sports  41.360001  28.959999      3.77         8.45          76.0   

In [136]:
df_actual = df_actual.sort_values(by = 'year_of_release')
print(df_actual)
#Отсортируем данные в порядке возрастания года

                                               name platform year_of_release  \
7300                      ms. pac-man: maze madness      N64      2000-01-01   
479                           the sims: livin large       PC      2000-01-01   
11463            derby tsuku: derby uma o tsukurou!       DC      2000-01-01   
10435              foxkids.com micro maniacs racing       PS      2000-01-01   
4406   walt disney world quest: magical racing tour       PS      2000-01-01   
...                                             ...      ...             ...   
2668                       aliens: colonial marines     X360      2013-01-01   
14932                             kamigami no asobi      PSP      2013-01-01   
12403                       the guided fate paradox      PS3      2013-01-01   
906                        mario party: island tour      3DS      2013-01-01   
15325                               urakata hakuoki      PSP      2013-01-01   

              genre  na_sales  eu_sales

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [139]:
def categorize_games_user_score (row):
    if 0 <= row['user_score'] < 3:
        return 'низкая оценка'
    elif 3 <= row['user_score'] < 8:
        return 'средняя оценка'
    elif 8 <= row['user_score'] <= 10:
        return 'высокая оценка'
    else: 'без оценки'
    
df_actual['user_score_category'] = df_actual.apply(categorize_games_user_score, axis = 1)
print(df_actual)

                                               name platform year_of_release  \
7300                      ms. pac-man: maze madness      N64      2000-01-01   
479                           the sims: livin large       PC      2000-01-01   
11463            derby tsuku: derby uma o tsukurou!       DC      2000-01-01   
10435              foxkids.com micro maniacs racing       PS      2000-01-01   
4406   walt disney world quest: magical racing tour       PS      2000-01-01   
...                                             ...      ...             ...   
2668                       aliens: colonial marines     X360      2013-01-01   
14932                             kamigami no asobi      PSP      2013-01-01   
12403                       the guided fate paradox      PS3      2013-01-01   
906                        mario party: island tour      3DS      2013-01-01   
15325                               urakata hakuoki      PSP      2013-01-01   

              genre  na_sales  eu_sales

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [140]:
def categorize_games_critic_score (row):
    if 0 <= row['critic_score'] < 30:
        return 'низкая оценка'
    elif 30 <= row['critic_score'] < 80:
        return 'средняя оценка'
    elif 80 <= row['critic_score'] <= 100:
        return 'высокая оценка'
    else: 'без оценки'
    
df_actual['critic_score_category'] = df_actual.apply(categorize_games_critic_score, axis = 1)
print(df_actual)

                                               name platform year_of_release  \
7300                      ms. pac-man: maze madness      N64      2000-01-01   
479                           the sims: livin large       PC      2000-01-01   
11463            derby tsuku: derby uma o tsukurou!       DC      2000-01-01   
10435              foxkids.com micro maniacs racing       PS      2000-01-01   
4406   walt disney world quest: magical racing tour       PS      2000-01-01   
...                                             ...      ...             ...   
2668                       aliens: colonial marines     X360      2013-01-01   
14932                             kamigami no asobi      PSP      2013-01-01   
12403                       the guided fate paradox      PS3      2013-01-01   
906                        mario party: island tour      3DS      2013-01-01   
15325                               urakata hakuoki      PSP      2013-01-01   

              genre  na_sales  eu_sales

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [142]:
groupde_data_user_score = df_actual.groupby('user_score_category')['name'].count()
groupde_data_user_score = groupde_data_user_score.sort_values(ascending= False)
print(groupde_data_user_score)

user_score_category
низкая оценка     4859
средняя оценка    2270
высокая оценка    1575
Name: name, dtype: int64


In [110]:
groupde_data_critic_score = df_actual.groupby('critic_score_category')['name'].count()
groupde_data_critic_score = groupde_data_critic_score.sort_values(ascending= False)
print(groupde_data_critic_score)

critic_score_category
низкая оценка     4412
средняя оценка    3280
высокая оценка    1012
Name: name, dtype: int64


- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [100]:
popular_platform = df.groupby('platform')['name'].count()
popular_platform = popular_platform.sort_values(ascending= False)
print(popular_platform.head(7))

platform
PS2    1865
DS     1791
PS     1126
WII     937
PSP     905
PS3     693
GBA     623
Name: name, dtype: int64


---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

### 1. Топ 7 популярных платформ:
 - консольные платформы были самыми популярными 
 - топ 1 занимает платформа PS2 со значением 1865 выпущенных игр
 - в топ 7 4 позиции занимает плтаформы PS, остальные три платформы Nintendo
### 2. Категории игр по оценкам пользователей за период с 2000 по 2013 год: 
 - по полученным данным можно сказть, что больше всего игр в категории 'низкая оценка', скорее всего это связанно с тем, что в этот период была большая конкуренция по играм различных жанров. В то время был всплеск игровой индустрии и естественно многие проеты были неудачными.
### 3. Категории игр по оценкам критиков за период с 2000 по 2013 год: 
 - результат +- такой же как и в оценке пользователей, ни низких оценок больше, а высоких меньше в среднем на 30%, но градация значений ктегорий зекральна.

Так же хочу добавить, что очень много пропусков в столбцке rating, где указаны присвоения возростной категории. Хоть в нашем анализе эти знаяения не принимали участия, тем не менее это очень важный факт, который нужно учесть пр подготовке статьи. У топовых игр такая оценка точно должна быть, следует либо дорабоать данные. либо удалить строки, где нет этой оценки. 
 Так же в этой базе данных очень много столбцов имели некорретный тип данных, это нужно учесть при дальнейшей работе с ней.